src-input/duk_bi_encoding.c

```diff
@@ -9,91 +9,103 @@
 
 #if defined(DUK_USE_ENCODING_BUILTINS)
 typedef struct {
-	duk_uint8_t *out;      /* where to write next byte(s) */
-	duk_codepoint_t lead;  /* lead surrogate */
+	duk_uint8_t *out;       /* where to write next byte(s) */
+	duk_codepoint_t lead;   /* lead surrogate */
 } duk__encode_context;
 
 typedef struct {
-	duk_bool_t bom_seen;
-	duk_bool_t fatal;
-	duk_bool_t ignore_bom;
-
 	/* UTF-8 decoding state */
-	duk_codepoint_t codepoint;  /* built up incrementally */
-	duk_int_t needed;           /* how many more bytes we need */
-	duk_uint8_t upper;          /* max value of next byte (decode error otherwise) */
-	duk_uint8_t lower;          /* min value of next byte (ditto) */
+	duk_codepoint_t codepoint;   /* built up incrementally */
+	duk_uint8_t upper;           /* max value of next byte (decode error otherwise) */
+	duk_uint8_t lower;           /* min value of next byte (ditto) */
+	duk_uint8_t needed;          /* how many more bytes we need */
+	duk_uint8_t bom_handled;     /* BOM seen or no longer expected */
+
+	/* Decoder configuration */
+	duk_uint8_t fatal;
+	duk_uint8_t ignore_bom;
 } duk__decode_context;
 
-#define DUK__U8_CONTINUE   0  /* continue to next byte */
-#define DUK__U8_CODEPOINT  1  /* have codepoint */
-#define DUK__U8_ERROR      2  /* decoding error */
-#define DUK__U8_RETRY      3  /* retry last byte, constraint: RETRY > ERROR */
+/* The signed duk_codepoint_t type is used to signal a decoded codepoint
+ * (>= 0) or various other states using negative values.
+ */
+#define DUK__CP_CONTINUE   (-1)  /* continue to next byte, no completed codepoint */
+#define DUK__CP_ERROR      (-2)  /* decoding error */
+#define DUK__CP_RETRY      (-3)  /* decoding error; retry last byte, constraint: RETRY > ERROR */
```

Comment:
I was a bit on the fence about this change: it doesn't affect performance on modern processors (but probably does improve a little bit on ARM etc), but it does reduce footprint a bit (total footprint reduction is around 350 bytes now on x64), so it seemed to make sense.

But assuming that the same code would be used to handle CESU-8 decoding too, which would be good for footprint, it would make sense to revert this particular change. Otherwise two decoders or some other hackery would be required which might increase footprint overall.

For performance oriented builds the decoding should be table oriented anyway (and thus handled by different code) so IMO the master baseline should be as footprint friendly as possible, and then have best performance possible within that constraint.